### PR TITLE
[plugin]: orama schema from json schema

### PR DESCRIPTION
**Pull request by ilteoood**
Possible solution for #445.

This is an initial basic solution, just to understand if orama wants this kind of plugin.

Todo for the future: support `$ref`

### CODE REVIEW

File: packages/schema/src/jsonSchema.ts [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    return ORAMA_SUPPORTED_TYPES.has(jsonSchema.type) || isArraySupportedByOrama(jsonSchema);
+}
+
+const extractOramaType = (jsonSchema: JSONSchema4): SearchableType => {
+    const oramaType = ORAMA_SUPPORTED_TYPES.has(jsonSchema.type) ? jsonSchema.type : `${(jsonSchema.items as JSONSchema4)!.type}[]`
+
+    return oramaType as SearchableType
+}
+
+export const schemaFromJson = (jsonSchema: JSONSchema4) => {
```
Comment (allevo) [inline packages/schema/src/jsonSchema.ts:31]: For now, the $ref is not supported, and it is ok. Anyway, if someone in the future wants to implement it, we should allow async operation. For instance, we need to allow the download of external references.
I prefer to move this to async function.

**Review by allevo — COMMENTED** [review]
Amazing!
Thanks, this is very helpful!

Only one note left

File: packages/schema/package.json [hunk 2]
```diff
@@ -0,0 +1,57 @@
+{
+  "name": "@orama/schema",
```
Comment (allevo) [inline packages/schema/package.json:2]: @orama/plugin-json-schema

File: packages/schema/src/jsonSchema.ts [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    assertTypeObject(jsonSchema)
+
+    const oramaSchema: Schema = {}
+
+    for (const [propertyName, propertyDefinition] of Object.entries(jsonSchema.properties || {})) {
+        if (isSupportedByOrama(propertyDefinition)) {
+            oramaSchema[propertyName] = extractOramaType(propertyDefinition)
+        } else if (isJsonObject(propertyDefinition)) {
+            oramaSchema[propertyName] = schemaFromJson(propertyDefinition)
+        }
```
Comment (allevo) [inline packages/schema/src/jsonSchema.ts:41]: should we throw an error here?

**Comment by ilteoood** [comment]
@allevo I'll be on holiday until next week, I'll address all of them once I'll be back!

File: packages/schema/package.json (hunk 2 above)
Comment (ilteoood) [reply packages/schema/package.json:2]: I was thinking to keep it generic so if in the future you would like to generate the schema from other sources, like yaml or toml, you don't have lots of packages to maintain

File: packages/schema/src/jsonSchema.ts (hunk 3 above)
Comment (ilteoood) [reply packages/schema/src/jsonSchema.ts:41]: Does orama throws if it receive a wrong schema? If not, I would not throw here either

Comment (ilteoood) [reply packages/schema/src/jsonSchema.ts:41]: Looking at the `validateSchema` function provided by Orama, it's ignoring the invalid types, I would do the same here

**Pull request closed without merging**